File: leetcode/sum-root-to-leaf-numbers.py
# 129. Sum Root to Leaf Numbers
# 🟠 Medium
#
# https://leetcode.com/problems/sum-root-to-leaf-numbers/
#
# Tags: Tree - Depth-First Search - Binary Tree

import timeit
from typing import Optional

from utils.binary_tree import TreeNode


# Traverse the tree using preorder DFS, add the value of the current
# node to the path sum, if the node is a leaf, return that value,
# otherwise call the function for any non-null children.
#
# Time complexity: O(n) - We will visit every node in the tree and do
# constant work for each.
# Space complexity: O(h) - The height of the call stack will be the
# height of the tree, worst O(n), best O(log(n)).
#
# Runtime 28 ms Beats 88.93%
# Memory 13.8 MB Beats 95.89%
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        def dfs(node: TreeNode, path: int) -> int:
            path *= 10
            path += node.val
            if not node.left and not node.right:
                return path
            return (dfs(node.left, path) if node.left else 0) + (
                dfs(node.right, path) if node.right else 0
            )

        return dfs(root, 0)


def test():
    executors = [Solution]
    tests = []
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.methodCall(t[0])
                exp = t[1]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/sum-root-to-leaf-numbers.rs
// 129. Sum Root to Leaf Numbers
// 🟠 Medium
//
// https://leetcode.com/problems/sum-root-to-leaf-numbers/
//
// Tags: Tree - Depth-First Search - Binary Tree

use std::cell::RefCell;
use std::rc::Rc;

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

struct Solution;
impl Solution {
    // Traverse the tree using preorder DFS, add the value of the current node to the path sum, if
    // the node is a leaf, return that value, otherwise call the function for any non-null child.
    //
    // Time complexity: O(n) - We will visit every node in the tree and do constant work for each.
    // Space complexity: O(h) - The height of the call stack will be the height of the tree, worst
    // case O(n), best O(log(n)).
    //
    // Runtime 0 ms Beats 100%
    // Memory 2.1 MB Beats 44.44%
    #[allow(dead_code)]
    pub fn sum_numbers(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn dfs(node: Rc<RefCell<TreeNode>>, path_sum: i32) -> i32 {
            let path_sum = path_sum * 10 + node.borrow().val;
            match (node.borrow().left.clone(), node.borrow().right.clone()) {
                (None, None) => path_sum,
                (None, Some(child)) | (Some(child), None) => dfs(child, path_sum),
                (Some(left), Some(right)) => dfs(left, path_sum) + dfs(right, path_sum),
            }
        }
        match root {
            Some(root) => dfs(root, 0),
            None => 0,
        }
    }

    /// Same as the previous solution but use references and a match instead of cloning the Rc and
    /// using an if statement.
    ///
    /// Time complexity: O(n) - We visit every node and do O(1) work for each.
    /// Space complexity: O(h) - The call stack will grow to the size of the tree.
    ///
    /// Runtime 0 ms Beats 100%
    /// Memory 2.17 MB Beats 19%
    #[allow(dead_code)]
    pub fn sum_numbers_ref(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn dfs(opt: Option<&Rc<RefCell<TreeNode>>>, path_sum: i32) -> i32 {
            match opt {
                Some(rc) => {
                    let node = rc.borrow();
                    let path_sum = path_sum * 10 + node.val;
                    match (&node.left, &node.right) {
                        (None, None) => path_sum,
                        _ => dfs(node.left.as_ref(), path_sum) + dfs(node.right.as_ref(), path_sum),
                    }
                }
                None => 0,
            }
        }
        dfs(root.as_ref(), 0)
    }
}

// Tests.
fn main() {
    println!("\n\x1b[92m» No tests in this file\x1b[0m");
}
